man/systemd-repart.xml

```diff
@@ -500,6 +503,44 @@
         <xi:include href="version-info.xml" xpointer="v255"/></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--copy-source=</option><replaceable>PATH</replaceable></term>
```

Comment:
Is this actually useful? When building a DDI you're going to be using --definitions anyway so those can't be looked up in --root. That leaves the machine ID for the random seed which is avoided by specifying a seed explicitly and specifiers, which I don't think are a huge problem?

So I'm having trouble seeing the value this adds over just using --root

Reply:
I was thinking about this, but conceptually I was thinking that `--root=`/`--image=` really is about *comprehensively* working based on some other tree. But here we want something much more focussed: just specify a source tree to copy from, and not more. 

Right now --root= is about four things at least:

* CopyFiles=
* CopyBlocks=
* .conf definition files
* machine ID

For our usecase here we want exactly one of these, and we want to apply it all to some image file that is unrelated to the root tree entirely, hence it sounded a bit weird to me to "overload" --root= for our purpose here.

Moreover, I wouldn't discount that sooner or later we want more resources that --root= affects (for example TPM enrollment metadata such as pcrlock stuff, that is maintained inside of images), and it would be strange to each time then making sure it doesn't fuck up the --make-ddi= usecase.

Or to turn this around, I'd rather keep the door open to allow --root= *together* with --copy-source= one day, so that the config/seed/… and everything is read from the root dir, but the file tree to copy from is kept separate.

Does that make any sense?

Reply:
But currently it means that we take all that data from the rootfs when using --copy-source which might not be intended when building DDIs. So if you put up a PR next for the TPM enrollment metadata stuff we will take those from the host system when building a DDI. This seems like it will trip users up sooner or later. So having this default to using the stuff from the host does not seem like a good default to me

Reply:
To add to this: this code makes the DDI types implemented via subdirs in `/usr/lib/systemd/repart/definitions/`, to make this extensible by users one day. I purposefully didn't document the path and that people can juts define their own for now, but I definitely want to keep this open, for example so that people define their own flavours that use squashfs rather than erofs, or for confext that are locked to some TPM). But I think the `/usr/lib/systemd/repart/definitions/` dir should one day also be read relative to `--root=`, but once we allow that then --root= cannot be used as copy source anymore without things becoming very messy.

(For now I explicitly generate an error if --root= and --make-ddi= are used together, so that we can open this up eventually still without breaking any current uses).

Anyway, I hope this underlines why I think `--root=` should be separate from `--copy-source`: one is for everything and most importantly configration, while the other is really just about redirecting a single type of *content* resource.

Reply:
> But currently it means that we take all that data from the rootfs when using --copy-source which might not be intended when building DDIs. So if you put up a PR next for the TPM enrollment metadata stuff we will take those from the host system when building a DDI. This seems like it will trip users up sooner or later. So having this default to using the stuff from the host does not seem like a good default to me

you have a point. I think --make-ddi= should set things up so that we do not leak host identity into the DDIs generated here. i.e. machine ID should not be read and so on. will update.

Reply:
So maybe add --root support for this immediately, and require specifying --root when --make-ddi is used so that users knowingly opt-in to using the host filesystem as the source of truth for all the rest?

Reply:
--make-ddi= is supposed to just make things easy, I am not sure we should confuse people with that, and require them to type even longer command lines.